Neutrinos

sbird opened the pull request:
Merge a massive neutrino implementation (since I am here).

This is based on the kspace-neutrinos implementation of Ali-Haimoud & Bird 2012, which follows neutrinos analytically, but it also incorporates hybrid neutrinos, which are initially tracers and then later gravitate. It uses code from the kspace-neutrinos repo (which has its own tests), but does not use the generic interface layer from that repo. This should have little to no effect if neutrinos are massless.

sbird commented:
Note that does not include neutrino support in MP-GenIC - that piece requires different transfer functions, which I didn't implement yet.

rainwoodman commented:
But that means we can't test it .. Shall we add something real simple to genic (could be wrong) just to drive the tests -- e.g. having the initial neutrinos on a uniform grid (with no power, since you said it doesn't matter?)

I worry if we don't merge it becomes too divergent; but it is also not good if we merge it without being able to drive it

sbird commented:
I can add a basic implementation of multiple transfer functions to MP-GenIC - that let's us test the important bit, the semi-linear neutrinos, and it is useful for other reasons (testing ADAPTIVE_GRAVSOFT_FORGAS, for example). It is not too hard. I'll see what I can do and let's hold off the merge until then. Do you like the general approach?

rainwoodman commented:
Yes. your approach address my biggest concern!

andreufont commented:
Hi @sbird - As you know, I'm very interested in this pull request. Please, let me know when it is ready for me to try it :-)

rainwoodman commented:
This is a rather large PR. I need to digest it -- one possibile approach is I do the rebase over the weekend to force myself to understand it better. But we do still want a driver otherwise there is no way for people to even run it..

sbird commented:
Do you still want to do the rebase? It needs to be done so I can add the GenIC integration using the different transfer function support.

rainwoodman commented:
I won't have the time for it -- I'll work on the CI infrastructure on nersc today. So go ahead.

sbird commented:
That is a basic neutrino driver for GenIC - no support for the hybrid neutrinos yet. 
@andreufont , feel free to give this a try.

sbird commented:
Drivers for hybrid neutrinos are now implemented. This should work and is ready to be reviewed.

rainwoodman commented on allvars.h:
```diff
@@ -288,6 +288,14 @@ extern struct global_data_all_processes
+    double HybridVcrit; /*!< Critical velocity switching between particle
+                          and analytic solvers when hybrid neutrinos are on*/
+    double HybridNuPartTime; /*!< Redshift at which hybrid neutrinos switch on*/
+    char CAMBTransferFunction[100]; /*!< CAMB transfer function for initial neutrino power*/
```
100 might be too short.

sbird replied:
You're right, I will fix that.

rainwoodman commented on begrun.c:
```diff
@@ -213,13 +217,16 @@ set_units(void)
     message(0, "UnitEnergy_in_cgs = %g \n", All.UnitEnergy_in_cgs);
     message(0, "Photon density OmegaG = %g\n",All.CP.OmegaG);
-    message(0, "Massless Neutrino density OmegaNu0 = %g\n",All.CP.OmegaNu0);
+    if(!All.MassiveNuLinRespOn)
```
Perhaps we can start using '.' to create long parameter names. What is the relation between MassiveNuLinResp and HybridNeutrinos?

sbird replied:
You mean to organise the parameters for a module into a short struct, which is in turn a member of All? HybridNeutrinos is a sub-option of MassiveNuLinResp, indeed.

rainwoodman replied:
At least a first clean up is the parameter names, `StarFormation.Enabled, StarFormation.Wind.Enabled, StarFormation.Wind.Model StarFormation.Wind; And neutrino goes to Neutrino.Enabled, Neutrino.Hybrid.Enabled, ... Using structs (like Cosmology) in the internal storage will ease later sharing the parser code between GenIC and Gadget too.

sbird replied:
I like this idea but I wonder if we should do it all at once, in a separate (later) pull. It is kind of strange when only the neutrino parameter obey this convention. Also it would be nice to have some kind of support for these structs from the param reading code - you had some idea to embed lua for reading parameters at one point, didn't you?

rainwoodman replied:
OK. Let's plan for a massive parameter overhual later.

rainwoodman commented on begrun.c:
```diff
@@ -213,13 +217,16 @@ set_units(void)
     message(0, "Photon density OmegaG = %g\n",All.CP.OmegaG);
-    message(0, "Massless Neutrino density OmegaNu0 = %g\n",All.CP.OmegaNu0);
+    if(!All.MassiveNuLinRespOn)
+        message(0, "Massless Neutrino density OmegaNu0 = %g\n",get_omega_nu(&All.CP.ONu, 1));
```
In class they call this omega_ur, ultra relativistic species. A reason it is cleaner is because it is hard to distinguish nu,  massive nu and masslessnu. What convention are we using here? nu is massless and mnu is massive?

sbird replied:
The convention here is that all neutrino masses are computed by get_omega_nu, whether massless or massive. There is no split between massless and massive neutrinos at all. MassiveNuLinRespOn controls only whether the neutrinos are supposed to be a component of Omega0.

rainwoodman commented on begrun.c:
```diff
@@ -213,13 +217,16 @@ set_units(void)
         /* note that this value is inaccurate if there is massive neutrino. */
+        double OmegaTot = All.CP.OmegaG + All.CP.OmegaK + All.CP.Omega0 + All.CP.OmegaLambda;
+        if(All.CP.MNu[0] + All.CP.MNu[1] + All.CP.MNu[2] == 0)
+            OmegaTot += get_omega_nu(&All.CP.ONu, 1);
```
ONu and MNu are not independent parameters -- Onu depends on the number of massless species in MNu. From this perspective this line here is confusing -- what if I only have 1 massless species?

sbird replied:
Yes - this is a wrong condition. It should read 
if(!All.MassiveNuLinRespOn)
OmegaTot += get_omega_nu()
Apologies, I wrote this late last night.

rainwoodman commented on cosmology.c:
```diff
@@ -33,17 +33,11 @@ void init_cosmology(Cosmology * CP_in)
-     * but we absorbed N_eff into T_nu above. */
-    CP->OmegaNu0 = CP->OmegaG * 7. / 8 * pow(TNu0_TCMB0, 4) * 3;
+    init_omega_nu(&CP->ONu, CP->MNu, TimeBegin, CP->HubbleParam, CP->CMBTemperature);
+    /* Neutrinos will be included in Omega0, if massive.
```
This convention is inconvenient. I remember redshift dependency of massive neutrino is not a**3 -- due to the distribution function. So If you include Oncdm in Omega0, then special care has to be taken for the redshift dependency...

rainwoodman replied:
We may want to drastically change the input parameters such that it takes a proper set (CLASS set) of parameters instead of the crippled Omega0.

rainwoodman commented on cosmology.c:
```diff
@@ -57,15 +51,12 @@ double hubble_function(double a)
-        /* massless neutrinos are added only if there is no (massive) neutrino particle.*/
-        if(CP->MasslessNeutrinosOn)
-            hubble_a += CP->OmegaNu0 / (a * a * a * a);
+        hubble_a += get_omega_nu(&CP->ONu, a);
```
What about the massive neutrino?

rainwoodman replied:
It currently is not in OmegaCDM.

rainwoodman replied:
if get_omega_nu returns the sum of massive and massless nu, this line looks even stranger ... aren't you supposed to divide this by `a**3` or `a**4`?

sbird replied:
No, it returns the sum of the massive neutrinos as a redshift dependent property: OmegaNu(a). For high masses we have OmegaNu(a) ~ MNu/(93.15 h^2) and for low masses OmegaNu(a) ~a^-4. Because the neutrino mass interpolates between the two (noticeably so) it seemed the least confusing thing to do

rainwoodman replied:
This touches how k-space-neutrino is implemented and I don't think it can be addressed in this PR; especially given that you are also embedding it in other places..

If you think of neutrino as a single fluid then yes, I see how the current get_omega_nu is the least confusing approach. massive neutrino still has a pressure component that scales like a^-4. I think that's likely what you mean by the neutrino mass interpolates between the two. 

However, after studying class somewhat carefully, I tend to agree with class where it divides neutrinos into components in terms of how they scale with a: ur (a**-4), mass-like (a**-3), pressure-like (a**-4), like here:

https://github.com/lesgourg/class_public/blob/master/source/background.c#L354

sbird replied:
I think the reason I like the single fluid is that it matches how I am treating the perturbations, and it matches well with how particles are treated in neutrino particle simulations: they are a single fluid, described by particles. It also has a neat interpretation in terms of the maximum neutrino velocity for the hybrid particles: you can set Max_nuvel = C/4 (say) and then the relativistic component is taken care of automatically.

rainwoodman replied:
I see. The hybrid particles deals with mass-like and pressure-like massive neutrinos the same time. What about the ultra relativistic species?

sbird replied:
Not sure what you mean exactly? pressure-like massive neutrinos are the ultra-relativistic species?

rainwoodman replied:
No. I mean there is a pressure-like component to the overall density due to massive neutrino that also scales like a**-4, in this line:
https://github.com/lesgourg/class_public/blob/master/source/background.c#L380

ultra-relativstic is already a**-4, at:
https://github.com/lesgourg/class_public/blob/master/source/background.c#L418

sbird replied:
Ah, I see. So you are proposing adding support for an arbitrary amount of ultra-relativistic species? 

At present we have OmegaG, which is so called because it is actually photons (and the parameter we set is the CMBTemperature). I think just renaming it OmegaUR would be not the right thing to do, because it is set up to be actual photons. But I think you are suggesting adding an extra species that redshifts like a^-4 and could map onto, for example, dark photons?

rainwoodman commented on forcetree.c:
```diff
@@ -753,7 +753,10 @@ force_update_node_recursive(int no, int sib, int tail, const struct TreeBuilder
+                /*Hybrid particle neutrinos do not gravitate at early times.
+                 * So do not add their masses to the node*/
+                if(!All.HybridNeutrinosOn || All.Time > All.HybridNuPartTime || P[next].Type != All.FastParticleType)
+                    add_particle_moment_to_node(&Nodes[no], &P[next]);
```
It might make more sense to move this if into add_particle_momentum_to_node.

rainwoodman commented on genic/main.c:
```diff
@@ -29,13 +30,22 @@ int main(int argc, char **argv)
+  const double shift_gas = -ProduceGas * 0.5 * (CP.Omega0 - CP.OmegaBaryon) / CP.Omega0 * meanspacing;
+  double shift_dm = ProduceGas * 0.5 * CP.OmegaBaryon / CP.Omega0 * meanspacing;
+  double shift_nu = 0;
+  if(!ProduceGas && NGridNu > 0) {
```
Does this suggest ProduceGas and NGridNu cannot be used the same time?

sbird replied:
No, that's fine. In that gas and DM are shifted, and NU sits in the middle on the grid points. As long as they do not initially have the same positions it is ok.

rainwoodman commented on genic/params.c:
```diff
@@ -32,12 +32,17 @@ create_parameters()
     param_declare_double(ps, "Omega_fld", OPTIONAL, 0, "Energy density of dark energy fluid.");
     param_declare_double(ps, "w0_fld", OPTIONAL, -1., "Dark energy equation of state");
     param_declare_double(ps, "wa_fld", OPTIONAL, 0, "Dark energy evolution parameter");
+    param_declare_double(ps, "MNue", OPTIONAL, 0, "First neutrino mass in eV.");
```
What about making this a list? we shall probably add param_declare_double_array.. Let's defer it to a new PR.

sbird replied:
Ok - that is a good idea but indeed let us defer it.

rainwoodman commented on genic/main.c:
```diff
@@ -46,19 +56,44 @@ int main(int argc, char **argv)
+      message(0,"F-D velocity scale: %g. Max particle vel: %g. Fraction of mass in particles: %g\n",v_th*sqrt(InitTime), Max_nuvel*sqrt(InitTime), total_frac);
+      setup_grid(shift_nu, TotNu);
+      for(i = 0; i< TotNu; i++)
+          add_thermal_speeds(P[i].Vel);
```
Is this Peculiar (a dx /dt ) or a**2 dx / dt

sbird replied:
Peculiar a dx/dt, I think.

rainwoodman replied:
You have to be careful, because the unit of P[i].Vel depends on whether UsePecuiliar Velocity is set.

sbird replied:
Well noticed - I had not accounted for that. Should be fixed now.

rainwoodman commented on gravpm.c:
```diff
@@ -265,6 +294,86 @@ static double sinc_unnormed(double x) {
+    /*Get delta_cdm_curr , which is P(k)^1/2, and skip bins with zero modes:*/
+    int nk_nonzero = 0;
+    for(i=0;i<PowerSpectrum.size;i++){
+        if (PowerSpectrum.Nmodes[i] == 0)
```
I think we shall eliminate these Nmodes[i] == 0 modes in the powerspectrum code -- here is where the linear binning works better -- at least all modes are non-empty!

sbird replied:
The linear binning actually produces a noticeably less accurate neutrino power spectrum, because of the excess smoothing induced (I checked this).

rainwoodman replied:
excessive smoothing at low k? because log binning have excessive smoothing at high k.

I think what you really want is, probably something per mode at low-k but then averaged at high k.

sbird replied:
Excessive smoothing at low k, indeed. Excessive smoothing at high k is ok for the neutrinos (even physical!) I think what you suggest is exactly what I really want.

rainwoodman replied:
We've been talking about doing this as an improvement of the model you implemented here. (actually a student here was supposed to work on this). Perhaps getting him to look at this after it is merged?

sbird replied:
That would be great.

rainwoodman replied:
Ok. I'll start nudging him. It is relatively simple but will likely be a mixture of commits on K-space-neutrino and MP-Gadget as I see now.

rainwoodman commented on gravpm.c:
```diff
@@ -265,6 +294,86 @@ static double sinc_unnormed(double x) {
+        /*Broadcast the transfer table*/
+        MPI_Bcast(transfer_init.logk,2*transfer_init.NPowerTable,MPI_DOUBLE,0,MPI_COMM_WORLD);
+        /*Initialise delta_tot*/
+        delta_tot_init(&delta_tot_table, nk_nonzero, PowerSpectrum.k, PowerSpectrum.P, &transfer_init, All.Time);
```
It may be useful to put all k-space-neutrino functions under the same prefix. not now though.

rainwoodman commented on gravpm.c:
```diff
@@ -299,6 +408,25 @@ void powerspectrum_compute(const int64_t k2, const int kpos[3], pfft_complex * c
 }
 
+/*Just read the power spectrum, without changing the input value.*/
+static void readout_power_spectrum(int64_t k2, int kpos[3], pfft_complex *value) {
```
I have been reserving readout for configuration space resampling operations. What about call this measure_power_transfer?

rainwoodman commented on gravpm.c:
```diff
@@ -299,6 +408,25 @@ void powerspectrum_compute(const int64_t k2, const int kpos[3], pfft_complex * c
+        tmp = sinc_unnormed(tmp);
+        f *= 1. / (tmp * tmp);
+    }
+    powerspectrum_compute(k2, kpos, value, f);
```
powerspectrum_add_mode?

rainwoodman commented on petaio.c:
```diff
@@ -124,6 +125,106 @@ petaio_build_selection(int * selection,
+        dims[0] = nk*ia;
+    ptrdiff_t strides[2] = {8, 8};
+    big_array_init(&deltas, delta_tot, "=f8", 2, dims, strides);
+    petaio_save_block(bf, "Neutrino/deltas", &deltas);
```
Perhaps saving k as well?And using captialized column names, Delta, K?

rainwoodman replied:
Looks like you can set dims[2] to {nk, ia} -- the result will read into python as a 2-d table.

sbird replied:
Is it ok to do that and read it back into a flat memory block in C?

rainwoodman replied:
It is still a flat memory block as long as the strides are correctly set.. I think you want
```
strides[2] = {8 * ia, 8}
dims[2] = {nk or 0, ia}
```
This will give you a tight representation of the data in C after reading.

sbird replied:
Thanks - does what I have now look right?

rainwoodman commented on petapm.h:
```diff
@@ -22,6 +22,12 @@ typedef struct {
+typedef struct {
+    void (*global_readout)(int64_t k2, int kpos[3], pfft_complex * value);
+    void (*global_analysis)(void);
+    void (*global_transfer)(int64_t k2, int kpos[3], pfft_complex * value);
```
May want to add arguments for the analysis result and parameter of the transfer.

sbird replied:
void * pointers to preserve the generic algorithm property?

sbird replied:
There is no-where for them to go - let me leave it as it is. It sets some structures which are effectively private members of the gravpm object. The other way I would have to pass the structures from petapm back to gravpm which would be a bit messy.

rainwoodman replied:
Sure. when we start to do the next round of global variables cleanups ...

sbird replied:
A never-ending task

rainwoodman commented:
There will be another 'minor' conflict after merging #143.

I have a few suggestions here and there but I don't recall very critical ones. 
I still think in the long run we will benefit from embedding CLASS to provide the cosmology and transfer functions. (dropping quite some code..) So if possible at all we shall use this chance to further massage our parametrizatoin of cosmology to align the with CLASS.

andreufont commented:
I've compiled this branch, and I'm able to run simple runs. Are there instructions / examples for how to specify neutrino masses, or this is not yet an option? The only difference that I see in the parameter files is in paramfile.gadget, where now we have:
MassiveNuLinRespOn = 0
What does this exactly do? Shouldn't we also have modifications in the initial conditions? 

Thanks!

sbird commented:
@andreufont : Look at examples/neutrinos, which should be an indicator of how to use it. You need to set MNue, MNum, MNut (although I will shortly make that a list) and MassiveNuLinRespOn = 1, both in  Gadget and in the ICs. There is documentation for the neutrino module also in kspace-neutrinos/README.

andreufont commented:
Thanks Simeon. The example runs (after fixing the parameter file for WindOn = 0). I'll play more with it tomorrow.

sbird commented:
Rebased, all but a couple of comments addressed and the paramfile updated.

sbird commented:
Ok, I have addressed all the comments I want to - as far as I'm concerned this is ready to merge. Let's wait a few days and see if @andreufont finds anything wrong with it?

rainwoodman commented:
We can also merge this first, then merge the fixes to master if andreu finds any. The bases for this consideration is that big PR tends to develop conflicts. If it is already merged we don't have to worry about these any more.

Another interesting dynamics we shall be aware is the submodule embedding of k-space-neutrino. It is the first 'science' module that we externally depend on. We only had utility modules previously.
Science modules actually representing the core business logic, but utility modules do not touch the core business. 

It is harder to insulate core business logic modules -- for example -- the choice of nu parametrization in k-space-neutrino dictates how nu shall be parametrized in MP-Gadget. Modifications of k-space-neutrino will have to propagate to other places, making it more costly. 

We will see how this plays out over time.

rainwoodman commented on petaio.c:
```diff
@@ -197,14 +198,15 @@ void petaio_read_neutrinos(BigFile * bf)
+    }
     big_array_init(&deltas, delta_tot, "=f8", 2, dims, strides);
-    petaio_read_block(bf, "Neutrino/deltas", &deltas, 1);
+    petaio_read_block(bf, "Neutrino/Deltas", &deltas, 1);
```
Is there a better name than Deltas? DeltaNuTotHistory?

sbird commented:
That is indeed a bit of a concern - if this becomes too difficult we could instead import the C files into MP-Gadget directly.

rainwoodman commented:
Conflict again!

sbird commented:
Fortunately a trivial one - maybe we should merge this before rebase errors start cropping up. If you are happy with the current state I will run one last long test with the current master and then we could merge?

rainwoodman commented:
no dark photons, yet. ur == massless neutrinos in CLASS convention. ncdm ==
massive neutrinos.

On Sun, Oct 29, 2017 at 5:47 PM, Simeon Bird <notifications@github.com>
wrote:

> *@sbird* commented on this pull request.
> ------------------------------
>
> In cosmology.c
> <https://github.com/rainwoodman/MP-Gadget/pull/124#discussion_r147602171>:
>
> >
>      if(CP->RadiationOn) {
>          hubble_a += CP->OmegaG / (a * a * a * a);
> -        /* massless neutrinos are added only if there is no (massive) neutrino particle.*/
> -        if(CP->MasslessNeutrinosOn)
> -            hubble_a += CP->OmegaNu0 / (a * a * a * a);
> +        hubble_a += get_omega_nu(&CP->ONu, a);
>
> Ah, I see. So you are proposing adding support for an arbitrary amount of
> ultra-relativistic species?
>
> At present we have OmegaG, which is so called because it is actually
> photons (and the parameter we set is the CMBTemperature). I think just
> renaming it OmegaUR would be not the right thing to do, because it is set
> up to be actual photons. But I think you are suggesting adding an extra
> species that redshifts like a^-4 and could map onto, for example, dark
> photons?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rainwoodman/MP-Gadget/pull/124#discussion_r147602171>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIbTNSMpSCfWwIMjBYeULN_FuVMbX8sks5sxRy0gaJpZM4PuRtt>
> .
>

sbird commented:
So class splits massive and massless neutrinos not for aesthetics but for internal reasons related to the approximations made by Boltzmann solver; massless neutrino transfer functions are basically free for class to compute (so much so that camb computes them even if there are no massless neutrinos), whereas massive neutrinos are one of the most expensive things it does.

The advantage of a single variable for the neutrinos is that physically they are the same species, so splitting them into a massive and massless part is a bit arbitrary.
